etc/kayobe/kolla/config/prometheus/prometheus.yml.d/60-redfish.yml

```diff
@@ -0,0 +1,27 @@
+# yamllint disable-file
+---
+{% if seed_redfish_exporter_container_enabled | bool %}
+scrape_configs:
+  - job_name: redfish-exporter-seed
+    metrics_path: /redfish
+    scrape_timeout: 120s
+    scrape_interval: {{ 8 * groups['redfish_exporter_targets'] | length }}s
+    relabel_configs:
+     - source_labels: [__address__]
+       target_label: __param_target
+     - source_labels: [__param_target]
+       target_label: instance
+     - target_label: __address__
+       replacement: "{{ lookup('vars', admin_oc_net_name ~ '_ips')[groups.seed.0] }}:9610"
+    static_configs:
+{% for host in groups.get('redfish_exporter_targets', []) %}
+{% if hostvars[host]["redfish_exporter_scrape_group"] | default('overcloud') == 'overcloud' %}
```

Comment:
One issue I've noticed is that sometimes I get scrapes failing because the idrac is overloaded when multiple prometheus instances scrape the data at the same time. I had a few ideas:

- Shuffle the targets so the order was randomised on each host (easiest)
  - Use a constant seed for each host so that order is maintained i.e I want to stop config diff from showing a change each run.
- Add caching layer (possibly with haproxy) that would serve stale data whilst a scrape was in progress i.e only allow one scrape at a time. No idea if haproxy can do this, but other caching proxies can.

Reply:
See something like: https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_cache_lock

Reply:
Would a simple mutex/lock based on the BMC address in the exporter work?

Reply:
> Would a simple mutex/lock based on the BMC address in the exporter work?

That sounds simpler than the cache. I must admit I was assuming that we didn't change the exporter. I guess we may have to bump the scrape timeout to account for the queued scrapes.  Though I was possibly slightly off in my earlier analysis of simultaneous scrapes... I found out it wasn't reusing sessions, so we were basically exhausting all the available sessions. On the idracs I am using there is a maximum session count of 8 with a timeout of 1800s. If scraping at 600s you basically use three sessions per prometheus. This is what you see in the logs:

```
t: 503: {\"error\":{\"@Message.ExtendedInfo\":[{\"Message\":\"The maximum number of user sessions is reached.\",\"MessageArgs\":[],\"MessageArgs@odata.count\":0,\"MessageId\":\"IDRAC.2.8.RAC0218\",\"RelatedProperties\":[],\"RelatedProperties@odata.count\":0,\"Resolution\":\"Make sure that the number of active sessions is less than or equal to the threshold limit. To end a current session and start a new session, do one of the following: 1) Log out of iDRAC by using a Graphical User Interface (GUI). 2) End any unused session by running the following RACADM command at the Command Line Interface (CLI): racadm closessn -i [session id].Warning: Ending a session may abruptly stop any running operation on that specific iDRAC session.\",\"Severity\":\"Informational\"}],\"code\":\"Base.1.12.GeneralError\",\"message\":\"A general error has occurred. See ExtendedInfo for more information\"}}" source="redfish_collector.go:48"
```
Adjusting the session timeout also affects the webui so it is not desirable to tweak that setting. The number of sessions is non-configurable.

Reply:
So I need to raise an upstream bug as https://github.com/jenningsloy318/redfish_exporter/blob/d388474c4024d09df0f3ab0290068e9072a7cf21/collector/redfish_collector.go#L84 seems to be failing to end the session